src/main/java/com/iota/iri/network/pipeline/ReceivedStage.java

```diff
@@ -0,0 +1,80 @@
+package com.iota.iri.network.pipeline;
+
+import com.iota.iri.TransactionValidator;
+import com.iota.iri.controllers.TransactionViewModel;
+import com.iota.iri.network.neighbor.Neighbor;
+import com.iota.iri.service.snapshot.SnapshotProvider;
+import com.iota.iri.storage.Tangle;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Optional;
+
+/**
+ * The {@link ReceivedStage} stores the given transaction in the database, updates the arrival time and sender and then
+ * submits to the {@link BroadcastStage}.
+ */
+public class ReceivedStage {
+
+    private static final Logger log = LoggerFactory.getLogger(ReceivedStage.class);
+
+    private Tangle tangle;
+    private TransactionValidator txValidator;
+    private SnapshotProvider snapshotProvider;
+
+    /**
+     * Creates a new {@link ReceivedStage}.
+     * 
+     * @param tangle           The {@link Tangle} database used to store/update the transaction
+     * @param txValidator      The {@link TransactionValidator} used to store/update the transaction
+     * @param snapshotProvider The {@link SnapshotProvider} used to store/update the transaction
+     */
+    public ReceivedStage(Tangle tangle, TransactionValidator txValidator, SnapshotProvider snapshotProvider) {
+        this.txValidator = txValidator;
+        this.tangle = tangle;
+        this.snapshotProvider = snapshotProvider;
+    }
+
+    /**
+     * Stores the given transaction in the database, updates it status
+     * ({@link TransactionValidator#updateStatus(TransactionViewModel)}) and updates the sender.
+     * 
+     * @param ctx the received stage {@link ProcessingContext}
+     * @return a {@link ProcessingContext} which redirects to the {@link BroadcastStage}
+     */
+    public ProcessingContext process(ProcessingContext ctx) {
+        ReceivedPayload payload = (ReceivedPayload) ctx.getPayload();
+        Optional<Neighbor> optNeighbor = payload.getNeighbor();
+        TransactionViewModel tvm = payload.getTransactionViewModel();
+
+        boolean stored;
+        try {
+            stored = tvm.store(tangle, snapshotProvider.getInitialSnapshot());
+        } catch (Exception e) {
+            log.error("error persisting newly received tx", e);
+            optNeighbor.ifPresent(neighbor -> neighbor.getMetrics().incrInvalidTransactionsCount());
+            ctx.setNextStage(TransactionProcessingPipeline.Stage.ABORT);
+            return ctx;
+        }
+
+        if (stored) {
+            tvm.setArrivalTime(System.currentTimeMillis());
+            try {
+                txValidator.updateStatus(tvm);
+                // neighbor might be null because tx came from a broadcastTransaction command
```

Comment:
Yes, I've also mentioned it in the first comment from this PR. Since a transaction is not broadcasted back to the origin neighbor, we need to store it in our database. This is actually an enhancement because it doesn't make sense to not store it if you are supposed to broadcast it.

Reply:
Ok, I didn't have it for some reason in the notes I am taking